File: 0030 - Substring with Concatenation of All Words/solution.cpp
class Solution {
public:
    vector<int> findSubstring(string s, vector<string>& A) {
        int l = s.size(), m = A.size(), n = A[0].size();
        vector<int> res;
        unordered_map<string, int> W;
        for(auto& w : A) W[w]++;
        for(int i = 0; i < n; i++){
            int left = i;
            unordered_map<string, int> t;
            for(int j = i; j <= l-n; j += n){
                string w = s.substr(j, n);
                if(!W.count(w)) {t.clear(); left = j+n; continue;}
                t[w]++;
                while(left<=j && t[w]>W[w]){
                    t[s.substr(left, n)]--;
                    left += n;
                }
                if(j+n-left == m*n) res.push_back(left);
            }
        }
        return res;
    }
};

File: 0030 - Substring with Concatenation of All Words/solution.py
from collections import Counter
class Solution:
    def findSubstring(self, s: str, A: List[str]) -> List[int]:
        l, m, n, res, A = len(s), len(A), len(A[0]), [], Counter(A)
        for i in range(n):
            left, t = i, {}
            for j in range(i, l-n+1, n):
                w = s[j:j+n]
                if w not in A:
                    left, t = j+n, {}
                    continue
                t[w] = t.get(w, 0)+1
                while left <= j and t[w] > A[w]:
                    t[s[left:left+n]] -= 1
                    left += n
                if j+n-left == m*n:
                    res.append(left)
        return res